charts/gardener/provider-local/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json

```diff
@@ -250,14 +250,14 @@
           "refId": "A"
         },
         {
-          "expr": "sum(kube_pod_container_resource_limits_memory_bytes{pod=~\"machine-controller-manager-(.+)\"}) by (pod)",
+          "expr": "sum(kube_pod_container_resource_limits{resource=\"memory\", unit=\"byte\", pod=~\"machine-controller-manager-(.+)\"}) by (pod)",
           "format": "time_series",
           "intervalFactor": 1,
           "legendFormat": "Limits ({{pod}})",
           "refId": "B"
         },
         {
-          "expr": "sum(kube_pod_container_resource_requests_memory_bytes{pod=~\"machine-controller-manager-(.+)\"}) by (pod)",
+          "expr": "sum(kube_pod_container_resource_requests{resource=\"memory\", unit=\"byte\", pod=~\"machine-controller-manager-(.+)\"}) by (pod)",
```

Comment:
I see that the same expr is present in the provider extensions:

https://github.com/gardener/gardener-extension-provider-aws/blob/84af99fb6b2754230dd3bb50e54e8f782eefd0f7/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json#L259-L265

What about them? Do we break them with the kube-state-metrics version update? What is the plan to update them?

Reply:
Thank you again! All the extensions have to be checked as well - I'll do so now and prepare PRs.

Reply:
there are many extension repositories:

`gardener-repos.sh`

```sh
#!/bin/bash
curl -s -H "Accept: application/vnd.github.v3+json" -H "Authorization: token $TOKEN" 'https://api.github.com/orgs/gardener/repos?per_page=100&page=1' | jq '.[].clone_url' -r
curl -s -H "Accept: application/vnd.github.v3+json" -H "Authorization: token $TOKEN" 'https://api.github.com/orgs/gardener/repos?per_page=100&page=2' | jq '.[].clone_url' -r
```

`export TOKEN=<token>` from https://github.com/settings/tokens

Assuming the naming convention of having the `gardener-extension` prefix in the repository name, I cloned and checked all the extensions:

```
for i in $(./gardener-repos.sh | grep gardener-extension); do echo $i; git clone $i; done
```

and the `kube_pod_container_resource_...` metrics are indeed used in a few extensions (the other removed metric names are not used). (They seem to contribute their version of the mcm dashboard.)

```
$ grep -r kube_pod_container_resource -l *
gardener-extension-provider-alicloud/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json
gardener-extension-provider-aws/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json
gardener-extension-provider-azure/charts/internal/seed-controlplane/charts/remedy-controller-azure/rc-monitoring-dashboard.json
gardener-extension-provider-azure/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json
gardener-extension-provider-equinix-metal/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json
gardener-extension-provider-gcp/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json
gardener-extension-provider-openstack/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json
gardener-extension-provider-vsphere/charts/internal/machine-controller-manager/seed/mcm-monitoring-dashboard.json
gardener-extension-shoot-cert-service/charts/internal/shoot-cert-management-seed/cert-dashboard.json
```

I can prepare PRs, but I'm not sure how could we roll out this change with the least disruption and with manageable effort.

Note that the dashboard changes are a "breaking" change (independent of the extension considerations above) in the sense that the historic values can not be queried under the new name, only the new data points will be shown on the dashboard after rolling out this change, as if we have lost some data points. We could address this with an OR statement in promql that we could remove later on, but I'm not sure if the historic metric values are worth it: recent values will be shown on the dashboard.

For the mcm dashboards that are contributed by the extensions, we could create a recording rule to store the new metrics also under the old name, knowing that the release schedule of the extensions is different from gardener. Otherwise, the affected queries would return nothing on these dashboards (even for recent data) as long as the extension upgrade is rolled out. Hence I think it is worth providing a mitigation for this incompatibility.

Reply:
Thanks for check and the explanations! I think I am fine with the breaking change for the extension dashboard - only 2 panels work work until the extensions are released with the adapted dashboard. And it will be still possible to query the corresponding information by using a plain query with the new metric, right? So, let's prepare the PRs for the extensions and proceed with them after we merge this one.
Do you have additional concerns?

Reply:
> I think I am fine with the breaking change for the extension dashboard

Just for the record: the mitigation to store the metrics with a recording rule under their old name will prevent temporarily breaking the dashboards that are contributed by the extensions. The mitigation can be removed when all the extensions have been released with their respective PR.

> And it will be still possible to query the corresponding information by using a plain query with the new metric, right?

Yes, that would be possible.

> So, let's prepare the PRs for the extensions

Done. :)

> Do you have additional concerns?

No. Thank you for the thorough review and catching the metrics that I somehow missed!